server/src/main/java/org/apache/druid/discovery/DruidLeaderClient.java

```diff
@@ -122,11 +122,21 @@ public void stop()
 
   /**
    * Make a Request object aimed at the leader. Throws IOException if the leader cannot be located.
+   *
+   * @param cached Uses cached leader if true, else uses the current leader
    */
-  public Request makeRequest(HttpMethod httpMethod, String urlPath) throws IOException
+  public Request makeRequest(HttpMethod httpMethod, String urlPath, boolean cached) throws IOException
   {
     Preconditions.checkState(lifecycleLock.awaitStarted(1, TimeUnit.MILLISECONDS));
-    return new Request(httpMethod, new URL(StringUtils.format("%s%s", getCurrentKnownLeader(true), urlPath)));
+    return new Request(httpMethod, new URL(StringUtils.format("%s%s", getCurrentKnownLeader(cached), urlPath)));
```

Comment:
Mentioning here that the "cache" parameter may not really be necessary, since `pickOneHost` appears to be reading cached state already from `serverDiscoverySelector` or `druidNodeDiscovery`. 

The other callers of makeRequest (with caching true) call DruidLeaderClient.go() which will retry with getCurrentKnownLeader(false) on failures, so they weren't affected by this stale leader issue.

Given that `makeRequest` is used by multiple callers though and we've already cut 0.13.0, I don't think we need to risk changing that behavior (i.e., does getCurrentKnownLeader really needs its own caching) for now though.

Reply:
Thanks for checking, agree it may be risky to remove the `cache` from `getCurrentKnownLeader`, can do that post 0.13.0.